File: prototype/src/connection.ts
/**
 * Pgvpd — Per-Connection Handler
 *
 * State machine managing the lifecycle of a single client connection:
 *
 *   WAIT_STARTUP ──→ Parse StartupMessage, extract tenant from username
 *        │              Rewrite username, connect to upstream
 *        ▼
 *   AUTHENTICATING ──→ Proxy auth messages bidirectionally
 *        │              Detect AuthenticationOk from server
 *        ▼
 *   POST_AUTH ──────→ Forward ParameterStatus, BackendKeyData to client
 *        │              When ReadyForQuery arrives: buffer it, move to INJECTING
 *        ▼
 *   INJECTING ──────→ Send SET commands to upstream
 *        │              Consume server response (CommandComplete, ReadyForQuery)
 *        │              Forward buffered ReadyForQuery to client
 *        ▼
 *   TRANSPARENT ────→ Bidirectional pipe — all traffic forwarded as-is
 */

import * as net from "node:net";
import {
  MessageFramer,
  parseStartupMessage,
  buildStartupMessage,
  buildQueryMessage,
  buildErrorResponse,
  isSSLRequest,
  isCancelRequest,
  isAuthOk,
  isReadyForQuery,
  isErrorResponse,
  extractErrorMessage,
  SSL_DENY,
  BackendMsg,
  type BackendMessage,
} from "./protocol.js";
import type { PgvpdConfig } from "./config.js";
import { log } from "./log.js";

enum State {
  WAIT_STARTUP = "WAIT_STARTUP",
  CONNECTING = "CONNECTING",
  AUTHENTICATING = "AUTHENTICATING",
  POST_AUTH = "POST_AUTH",
  INJECTING = "INJECTING",
  TRANSPARENT = "TRANSPARENT",
  CLOSED = "CLOSED",
}

export class Connection {
  private state: State = State.WAIT_STARTUP;
  private clientFramer = new MessageFramer();
  private serverFramer = new MessageFramer();
  private upstream: net.Socket | null = null;
  private contextValues: string[] = [];
  private actualUser: string | null = null;
  private isBypass = false;
  private bufferedReadyForQuery: Buffer | null = null;
  private connId: string;

  constructor(
    private client: net.Socket,
    private config: PgvpdConfig,
    connId: string,
  ) {
    this.connId = connId;

    client.on("data", (data: Buffer) => this.onClientData(data));
    client.on("error", (err) => this.handleError("client", err));
    client.on("close", () => this.cleanup("client closed"));
  }

  // ─── Client Data Handler ────────────────────────────────────────────────

  private onClientData(data: Buffer): void {
    try {
      switch (this.state) {
        case State.WAIT_STARTUP:
          this.handleStartupData(data);
          break;

        case State.AUTHENTICATING:
        case State.POST_AUTH:
          // Forward client auth messages to server
          this.upstream?.write(data);
          break;

        case State.INJECTING:
          // Client shouldn't be sending data yet (they haven't received
          // ReadyForQuery), but buffer it just in case
          this.clientFramer.push(data);
          break;

        case State.TRANSPARENT:
          // Direct pipe to upstream
          this.upstream?.write(data);
          break;

        default:
          break;
      }
    } catch (err) {
      this.handleError("client-handler", err as Error);
    }
  }

  // ─── Startup Phase ──────────────────────────────────────────────────────

  private handleStartupData(data: Buffer): void {
    this.clientFramer.push(data);
    const msg = this.clientFramer.nextStartupMessage();
    if (!msg) return; // need more data

    // Handle SSLRequest — deny and wait for the real StartupMessage
    if (isSSLRequest(msg)) {
      log.debug(this.connId, "SSL request denied");
      this.client.write(SSL_DENY);
      // Client will retry with a plaintext StartupMessage
      return; // stay in WAIT_STARTUP
    }

    // Handle CancelRequest — not supported in v0.1
    if (isCancelRequest(msg)) {
      log.debug(this.connId, "Cancel request — closing");
      this.client.end();
      this.state = State.CLOSED;
      return;
    }

    // Parse the StartupMessage
    const startup = parseStartupMessage(msg);
    const rawUser = startup.params.get("user");

    if (!rawUser) {
      this.sendErrorAndClose("FATAL", "08004", "No username in StartupMessage");
      return;
    }

    // Check for superuser bypass
    if (this.config.superuserBypass.includes(rawUser)) {
      log.info(this.connId, `Superuser bypass: ${rawUser}`);
      this.isBypass = true;
      this.connectUpstream(msg); // forward original message as-is
      return;
    }

    // Extract tenant context from username
    const sepIdx = rawUser.indexOf(this.config.tenantSeparator);
    if (sepIdx === -1) {
      this.sendErrorAndClose(
        "FATAL",
        "28000",
        `Username must contain context values separated by '${this.config.tenantSeparator}'`,
        `Expected format: role${this.config.tenantSeparator}value1 (e.g. app_user${this.config.tenantSeparator}acme)`,
      );
      return;
    }

    this.actualUser = rawUser.slice(0, sepIdx);
    const tenantPayload = rawUser.slice(sepIdx + 1);

    if (!this.actualUser || !tenantPayload) {
      this.sendErrorAndClose(
        "FATAL",
        "28000",
        "Empty role or context values in username",
      );
      return;
    }

    // Split payload into multiple values if multiple context variables configured
    if (this.config.contextVariables.length > 1) {
      this.contextValues = tenantPayload.split(this.config.valueSeparator);
    } else {
      this.contextValues = [tenantPayload];
    }

    if (this.contextValues.length !== this.config.contextVariables.length) {
      this.sendErrorAndClose(
        "FATAL",
        "28000",
        `Expected ${this.config.contextVariables.length} context value(s), got ${this.contextValues.length}`,
        `Context variables: ${this.config.contextVariables.join(", ")}. Separate values with '${this.config.valueSeparator}'.`,
      );
      return;
    }

    if (this.contextValues.some((v) => !v)) {
      this.sendErrorAndClose("FATAL", "28000", "Empty context value in username");
      return;
    }

    const contextSummary = this.config.contextVariables
      .map((v, i) => `${v}=${this.contextValues[i]}`)
      .join(", ");
    log.info(
      this.connId,
      `Context: ${contextSummary} | Role: ${this.actualUser} | DB: ${startup.params.get("database") || "default"}`,
    );

    // Rewrite the StartupMessage with the actual username
    const newParams = new Map(startup.params);
    newParams.set("user", this.actualUser);
    const rewrittenMsg = buildStartupMessage(newParams);

    this.connectUpstream(rewrittenMsg);
  }

  // ─── Upstream Connection ────────────────────────────────────────────────

  private connectUpstream(startupMsg: Buffer): void {
    this.state = State.CONNECTING;

    this.upstream = net.createConnection(
      {
        host: this.config.upstreamHost,
        port: this.config.upstreamPort,
      },
      () => {
        log.debug(
          this.connId,
          `Connected to upstream ${this.config.upstreamHost}:${this.config.upstreamPort}`,
        );

        // Send the (possibly rewritten) StartupMessage
        this.upstream!.write(startupMsg);

        // If any extra client data was buffered after the StartupMessage, forward it
        if (this.clientFramer.hasData()) {
          this.upstream!.write(this.clientFramer.drain());
        }

        this.state = this.isBypass ? State.TRANSPARENT : State.AUTHENTICATING;

        if (this.isBypass) {
          // For bypass connections, go straight to transparent pipe
          this.enterTransparent();
        }
      },
    );

    this.upstream.on("data", (data: Buffer) => this.onServerData(data));
    this.upstream.on("error", (err) => this.handleError("upstream", err));
    this.upstream.on("close", () => this.cleanup("upstream closed"));
  }

  // ─── Server Data Handler ────────────────────────────────────────────────

  private onServerData(data: Buffer): void {
    try {
      switch (this.state) {
        case State.TRANSPARENT:
          // Direct pipe to client
          this.client.write(data);
          break;

        case State.AUTHENTICATING:
          this.handleAuthData(data);
          break;

        case State.POST_AUTH:
          this.handlePostAuthData(data);
          break;

        case State.INJECTING:
          this.handleInjectionResponse(data);
          break;

        default:
          // Forward anyway as a safety net
          this.client.write(data);
          break;
      }
    } catch (err) {
      this.handleError("server-handler", err as Error);
    }
  }

  // ─── Authentication Phase ───────────────────────────────────────────────

  private handleAuthData(data: Buffer): void {
    this.serverFramer.push(data);

    let msg: BackendMessage | null;
    while ((msg = this.serverFramer.nextBackendMessage()) !== null) {
      if (isAuthOk(msg)) {
        log.debug(this.connId, "Authentication OK");
        this.client.write(msg.raw);
        this.state = State.POST_AUTH;

        // Process any remaining buffered messages in POST_AUTH mode
        if (this.serverFramer.hasData()) {
          this.handlePostAuthData(this.serverFramer.drain());
        }
        return;
      }

      if (isErrorResponse(msg)) {
        const errMsg = extractErrorMessage(msg);
        log.warn(this.connId, `Auth error from server: ${errMsg}`);
      }

      // Forward all auth messages to client (auth challenges, errors, etc.)
      this.client.write(msg.raw);
    }
  }

  // ─── Post-Auth Phase ────────────────────────────────────────────────────

  private handlePostAuthData(data: Buffer): void {
    this.serverFramer.push(data);

    let msg: BackendMessage | null;
    while ((msg = this.serverFramer.nextBackendMessage()) !== null) {
      if (isReadyForQuery(msg)) {
        // Buffer ReadyForQuery — don't send to client yet
        this.bufferedReadyForQuery = msg.raw;
        log.debug(this.connId, "ReadyForQuery buffered — injecting tenant context");
        this.injectTenantContext();
        return;
      }

      if (isErrorResponse(msg)) {
        const errMsg = extractErrorMessage(msg);
        log.warn(this.connId, `Post-auth error: ${errMsg}`);
      }

      // Forward ParameterStatus, BackendKeyData, NoticeResponse, etc.
      this.client.write(msg.raw);
    }
  }

  // ─── Context Injection ──────────────────────────────────────────────────

  private injectTenantContext(): void {
    this.state = State.INJECTING;

    // Build SET commands for each context variable.
    // Values are sanitized defensively even though they come from
    // the username field of the StartupMessage.
    const setClauses = this.config.contextVariables.map((varName, i) => {
      const safeValue = escapeLiteral(this.contextValues[i]);
      return `SET ${varName} = ${safeValue}`;
    });

    // Final SET ROLE enforces NOBYPASSRLS
    setClauses.push(`SET ROLE ${quoteIdent(this.actualUser!)}`);

    const sql = setClauses.join("; ") + ";";

    log.debug(this.connId, `Injecting: ${sql}`);
    this.upstream!.write(buildQueryMessage(sql));
  }

  private handleInjectionResponse(data: Buffer): void {
    this.serverFramer.push(data);

    let msg: BackendMessage | null;
    while ((msg = this.serverFramer.nextBackendMessage()) !== null) {
      if (isErrorResponse(msg)) {
        const errMsg = extractErrorMessage(msg);
        log.error(this.connId, `Context injection failed: ${errMsg}`);
        // Forward the error to the client and close
        this.client.write(msg.raw);
        this.cleanup("injection failed");
        return;
      }

      if (isReadyForQuery(msg)) {
        // Server has processed our SET commands.
        // Now send the buffered ReadyForQuery to the client.
        const summary = this.config.contextVariables
          .map((v, i) => `${v}=${this.contextValues[i]}`)
          .join(", ");
        log.info(this.connId, `Context set — ${summary}, role=${this.actualUser}`);

        if (this.bufferedReadyForQuery) {
          this.client.write(this.bufferedReadyForQuery);
          this.bufferedReadyForQuery = null;
        }

        // Flush any client data that arrived during injection
        if (this.clientFramer.hasData()) {
          this.upstream!.write(this.clientFramer.drain());
        }

        this.enterTransparent();
        return;
      }

      // CommandComplete ('C') for our SET commands — consume silently
      // ParameterStatus ('S') — forward to client (server may send these for SET)
      if (msg.type === BackendMsg.ParameterStatus) {
        this.client.write(msg.raw);
      }
      // Everything else from our injection (CommandComplete, etc.) — consume
    }
  }

  // ─── Transparent Pipe ───────────────────────────────────────────────────

  private enterTransparent(): void {
    this.state = State.TRANSPARENT;
    log.debug(this.connId, "Entering transparent pipe mode");

    // Flush any remaining buffered server data
    if (this.serverFramer.hasData()) {
      this.client.write(this.serverFramer.drain());
    }
  }

  // ─── Error Handling ─────────────────────────────────────────────────────

  private sendErrorAndClose(
    severity: string,
    sqlstate: string,
    message: string,
    detail?: string,
  ): void {
    log.warn(this.connId, `Rejecting: ${message}`);
    this.client.write(buildErrorResponse(severity, sqlstate, message, detail));
    this.client.end();
    this.state = State.CLOSED;
  }

  private handleError(source: string, err: Error): void {
    if ((err as NodeJS.ErrnoException).code === "ECONNRESET") {
      log.debug(this.connId, `${source} connection reset`);
    } else {
      log.error(this.connId, `${source} error: ${err.message}`);
    }
    this.cleanup(`${source} error`);
  }

  private cleanup(reason: string): void {
    if (this.state === State.CLOSED) return;
    this.state = State.CLOSED;

    log.debug(this.connId, `Closing: ${reason}`);

    this.client.destroy();
    this.upstream?.destroy();
    this.upstream = null;
  }
}

// ─── SQL Escaping ───────────────────────────────────────────────────────────

/**
 * Escape a string as a SQL literal (single-quoted, with ' doubled).
 * This is intentionally conservative — only allows alphanumeric, underscore,
 * hyphen, and dot characters. Everything else is rejected.
 */
function escapeLiteral(value: string): string {
  // Reject any characters that have no business being in a tenant ID
  if (!/^[a-zA-Z0-9_\-\.]+$/.test(value)) {
    throw new Error(`Invalid tenant ID: contains disallowed characters: ${value}`);
  }
  return `'${value.replace(/'/g, "''")}'`;
}

/**
 * Quote an identifier (double-quoted, with " doubled).
 * Same conservative character check.
 */
function quoteIdent(value: string): string {
  if (!/^[a-zA-Z0-9_]+$/.test(value)) {
    throw new Error(`Invalid identifier: ${value}`);
  }
  return `"${value.replace(/"/g, '""')}"`;
}

File: src/connection.rs
//! Per-Connection Handler
//!
//! Async state machine managing a single client connection through:
//!   WaitStartup → Authenticating → PostAuth → Injecting → Transparent

use bytes::BytesMut;
use std::sync::Arc;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tracing::{debug, error, info, warn};

use crate::config::Config;
use crate::protocol::{
    build_error_response, build_query_message, build_startup_message, escape_literal, quote_ident,
    try_read_backend_message, try_read_startup, StartupType, SSL_DENY,
};

/// Handle a single client connection through its full lifecycle.
pub async fn handle_connection(mut client: TcpStream, config: Arc<Config>, conn_id: u64) {
    let peer = client
        .peer_addr()
        .map(|a| a.to_string())
        .unwrap_or_else(|_| "unknown".into());
    debug!(conn_id, peer, "new connection");

    if let Err(e) = run_connection(&mut client, &config, conn_id).await {
        debug!(conn_id, error = %e, "connection ended");
    }
}

async fn run_connection(
    client: &mut TcpStream,
    config: &Config,
    conn_id: u64,
) -> Result<(), Box<dyn std::error::Error>> {
    // ─── Phase 1: Read StartupMessage ───────────────────────────────────

    let mut buf = BytesMut::with_capacity(1024);

    let startup = loop {
        client.read_buf(&mut buf).await?;

        match try_read_startup(&mut buf) {
            Some(StartupType::SslRequest) => {
                debug!(conn_id, "SSL request denied");
                client.write_all(SSL_DENY).await?;
                // Client will retry with plaintext StartupMessage
                continue;
            }
            Some(StartupType::CancelRequest) => {
                debug!(conn_id, "cancel request — closing");
                return Ok(());
            }
            Some(StartupType::Startup(s)) => break s,
            None => continue, // need more data
        }
    };

    let raw_user = startup.params.get("user").cloned().unwrap_or_default();
    if raw_user.is_empty() {
        send_error(client, "FATAL", "08004", "no username in StartupMessage").await;
        return Ok(());
    }

    let database = startup
        .params
        .get("database")
        .cloned()
        .unwrap_or_else(|| "default".into());

    // ─── Superuser bypass ───────────────────────────────────────────────

    if config.superuser_bypass.contains(&raw_user) {
        info!(conn_id, user = %raw_user, "superuser bypass");
        let mut server =
            TcpStream::connect((&*config.upstream_host, config.upstream_port)).await?;
        // Forward original startup message
        let original = build_startup_message(&startup.params);
        server.write_all(&original).await?;
        // Forward any buffered data
        if !buf.is_empty() {
            server.write_all(&buf).await?;
        }
        // Transparent pipe
        tokio::io::copy_bidirectional(client, &mut server).await?;
        return Ok(());
    }

    // ─── Extract tenant context from username ───────────────────────────

    let sep_idx = match raw_user.find(&config.tenant_separator) {
        Some(i) => i,
        None => {
            send_error(
                client,
                "FATAL",
                "28000",
                &format!(
                    "username must contain context values separated by '{}'",
                    config.tenant_separator
                ),
            )
            .await;
            return Ok(());
        }
    };

    let actual_user = &raw_user[..sep_idx];
    let tenant_payload = &raw_user[sep_idx + config.tenant_separator.len()..];

    if actual_user.is_empty() || tenant_payload.is_empty() {
        send_error(client, "FATAL", "28000", "empty role or context in username").await;
        return Ok(());
    }

    // Split payload into multiple values if needed
    let context_values: Vec<&str> = if config.context_variables.len() > 1 {
        tenant_payload.split(&config.value_separator).collect()
    } else {
        vec![tenant_payload]
    };

    if context_values.len() != config.context_variables.len() {
        send_error(
            client,
            "FATAL",
            "28000",
            &format!(
                "expected {} context value(s), got {}",
                config.context_variables.len(),
                context_values.len()
            ),
        )
        .await;
        return Ok(());
    }

    if context_values.iter().any(|v| v.is_empty()) {
        send_error(client, "FATAL", "28000", "empty context value in username").await;
        return Ok(());
    }

    let context_summary: String = config
        .context_variables
        .iter()
        .zip(context_values.iter())
        .map(|(var, val)| format!("{var}={val}"))
        .collect::<Vec<_>>()
        .join(", ");

    info!(
        conn_id,
        context = %context_summary,
        role = actual_user,
        database = %database,
        "tenant connection"
    );

    // ─── Phase 2: Connect to upstream ───────────────────────────────────

    let mut server = TcpStream::connect((&*config.upstream_host, config.upstream_port)).await?;
    debug!(
        conn_id,
        host = %config.upstream_host,
        port = config.upstream_port,
        "connected to upstream"
    );

    // Send rewritten StartupMessage
    let mut rewritten_params = startup.params.clone();
    rewritten_params.insert("user".into(), actual_user.to_string());
    let startup_msg = build_startup_message(&rewritten_params);
    server.write_all(&startup_msg).await?;

    // Forward any extra client data buffered after StartupMessage
    if !buf.is_empty() {
        server.write_all(&buf).await?;
        buf.clear();
    }

    // ─── Phase 3: Authentication ────────────────────────────────────────

    let mut server_buf = BytesMut::with_capacity(4096);
    let mut auth_done = false;

    while !auth_done {
        server.read_buf(&mut server_buf).await?;

        while let Some(msg) = try_read_backend_message(&mut server_buf) {
            if msg.is_auth_ok() {
                debug!(conn_id, "authentication OK");
                client.write_all(&msg.raw).await?;
                auth_done = true;
                break;
            }

            if msg.is_error_response() {
                warn!(conn_id, error = %msg.error_message(), "auth error from server");
            }

            // Forward auth challenges, errors, etc. to client
            client.write_all(&msg.raw).await?;

            // If server sent an auth challenge expecting a client response, relay it
            if msg.is_auth_challenge() {
                let mut client_buf = BytesMut::with_capacity(1024);
                client.read_buf(&mut client_buf).await?;
                server.write_all(&client_buf).await?;
            }
        }
    }

    // ─── Phase 4: Post-auth — wait for ReadyForQuery ────────────────────

    let buffered_ready: BytesMut = loop {
        // If there's leftover data in server_buf from auth phase, process it first
        if server_buf.is_empty() {
            server.read_buf(&mut server_buf).await?;
        }

        let mut ready_msg = None;
        while let Some(msg) = try_read_backend_message(&mut server_buf) {
            if msg.is_ready_for_query() {
                debug!(conn_id, "ReadyForQuery buffered — injecting context");
                ready_msg = Some(msg.raw);
                break;
            }

            if msg.is_error_response() {
                warn!(conn_id, error = %msg.error_message(), "post-auth error");
            }

            // Forward ParameterStatus, BackendKeyData, etc. to client
            client.write_all(&msg.raw).await?;
        }

        if let Some(raw) = ready_msg {
            break raw;
        }
    };

    // ─── Phase 5: Inject tenant context ─────────────────────────────────

    let mut set_clauses = Vec::new();
    for (var, val) in config.context_variables.iter().zip(context_values.iter()) {
        let safe_val = escape_literal(val)?;
        set_clauses.push(format!("SET {var} = {safe_val}"));
    }
    set_clauses.push(format!("SET ROLE {}", quote_ident(actual_user)?));
    let sql = set_clauses.join("; ") + ";";

    debug!(conn_id, sql = %sql, "injecting");
    let query_msg = build_query_message(&sql);
    server.write_all(&query_msg).await?;

    // Consume server's response to our SET commands
    loop {
        server.read_buf(&mut server_buf).await?;

        let mut injection_done = false;
        while let Some(msg) = try_read_backend_message(&mut server_buf) {
            if msg.is_error_response() {
                error!(conn_id, error = %msg.error_message(), "context injection failed");
                client.write_all(&msg.raw).await?;
                return Err(msg.error_message().into());
            }

            if msg.is_ready_for_query() {
                // Server confirmed our SETs. Send buffered ReadyForQuery to client.
                info!(
                    conn_id,
                    context = %context_summary,
                    role = actual_user,
                    "context set"
                );
                client.write_all(&buffered_ready).await?;
                injection_done = true;
                break;
            }

            // Forward ParameterStatus if server sends them for SET commands
            if msg.is_parameter_status() {
                client.write_all(&msg.raw).await?;
            }
            // CommandComplete — consume silently
        }

        if injection_done {
            break;
        }
    }

    // ─── Phase 6: Transparent pipe ──────────────────────────────────────

    debug!(conn_id, "transparent pipe");

    // Flush any remaining buffered server data
    if !server_buf.is_empty() {
        client.write_all(&server_buf).await?;
    }

    // Zero-copy bidirectional relay
    tokio::io::copy_bidirectional(client, &mut server).await?;

    Ok(())
}

async fn send_error(client: &mut TcpStream, severity: &str, sqlstate: &str, message: &str) {
    warn!(message, "rejecting connection");
    let msg = build_error_response(severity, sqlstate, message);
    let _ = client.write_all(&msg).await;
    let _ = client.shutdown().await;
}
